sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -354,7 +355,8 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
       query
     } else if (source == SOURCE_NAME_FOREACH) {
       assertNotPartitioned(SOURCE_NAME_FOREACH)
-      val sink = ForeachWriterTable[T](foreachWriter, ds.exprEnc)
+      val sink = ForeachWriterTable[Any](foreachWriter,
+        ds.exprEnc.asInstanceOf[ExpressionEncoder[Any]])
```

Comment:
Is this casting necessary? 
What if we replace 'Any' with '_' for `foreachWriter`?

Reply:
I tried making 
` private var foreachWriter: ForeachWriter[_] = null`

Then here if I do
```
val sink = ForeachWriterTable[_](foreachWriter, ds.exprEnc)
```
I get:
```
[error] /home/wei.liu/oss-spark/sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:358:37: unbound wildcard type
[error]       val sink = ForeachWriterTable[_](foreachWriter, ds.exprEnc)
[error]                                     ^
[error] one error found
```

If I do 
```
val sink = ForeachWriterTable[Any](foreachWriter, ds.exprEnc)
```
I get:
```
[error] /home/wei.liu/oss-spark/sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:358:36: overloaded method value apply with alternatives:
[error]   (writer: org.apache.spark.sql.ForeachWriter[Any],converter: Either[org.apache.spark.sql.catalyst.encoders.ExpressionEncoder[Any],org.apache.spark.sql.catalyst.InternalRow => Any])org.apache.spark.sql.execution.streaming.sources.ForeachWriterTable[Any] <and>
[error]   (writer: org.apache.spark.sql.ForeachWriter[Any],encoder: org.apache.spark.sql.catalyst.encoders.ExpressionEncoder[Any])org.apache.spark.sql.execution.streaming.sources.ForeachWriterTable[_]
[error]  cannot be applied to (org.apache.spark.sql.ForeachWriter[_$2], org.apache.spark.sql.catalyst.encoders.ExpressionEncoder[T])
[error]       val sink = ForeachWriterTable[Any](foreachWriter, ds.exprEnc)
[error]                                    ^
[error] one error found
```